jaxlib/mosaic/gpu/custom_call.cc

```diff
@@ -528,32 +577,109 @@ absl::StatusOr<std::tuple<void*, MosaicHostFunc*>> CachedCompileAndInit(
+  static auto &needs_nvshmem_barrier = *new absl::flat_hash_set<int64_t>;
+  return std::make_pair(&needs_nvshmem_barrier, &mutex);
+}
+
+absl::Status MosaicGpuExecute(gpuStream_t stream, ffi::RemainingArgs inputs,
+                              ffi::RemainingRets results,
+                              absl::string_view kernel_hash,
+                              absl::string_view module,
+                              bool use_custom_barrier, xla::RunId run_id) {
+  if (reinterpret_cast<const uintptr_t>(kernel_hash.data()) %
+          alignof(KernelHash) ||
+      kernel_hash.size() != sizeof(KernelHash)) {
     fprintf(stderr, "Misaligned opaque pointer\n");
     abort();
   }
-  auto hash = *reinterpret_cast<KernelHash*>(opaque);
+  auto hash = *reinterpret_cast<const KernelHash *>(kernel_hash.data());
   CUcontext ctx;
   if (cuCtxGetCurrent(&ctx) != CUDA_SUCCESS) {
     fprintf(stderr, "Failed to get current CUDA context\n");
     abort();
   }
   CacheKey key(hash, reinterpret_cast<uintptr_t>(ctx));
-  auto ctx_and_kernel = CachedCompileAndInit(key, opaque + sizeof(KernelHash));
-  if (!ctx_and_kernel.ok()) {
-    XlaCustomCallStatusSetFailure(status,
-                                  ctx_and_kernel.status().message().data(),
-                                  ctx_and_kernel.status().message().size());
-    return;
+  auto ctx_kernel_comm = CachedCompileAndInit(key, module.data());
+  if (!ctx_kernel_comm.ok()) {
+    return absl::InternalError(ctx_kernel_comm.status().message());
+  }
+  bool is_comm_used = std::get<2>(*ctx_kernel_comm);
+
+  std::vector<void *> buffers;
+  buffers.reserve(inputs.size() + results.size());
+  for (int i = 0; i < inputs.size(); ++i) {
+    buffers.push_back(inputs.get<ffi::AnyBuffer>(i).value().untyped_data());
+  }
+  for (int i = 0; i < results.size(); ++i) {
+    buffers.push_back(results.get<ffi::AnyBuffer>(i).value()->untyped_data());
+  }
+  void **buffers_ptr = buffers.data();
+  void *args[4] = {&std::get<0>(*ctx_kernel_comm), &stream, &buffers_ptr};
+
+  if (is_comm_used) {
+    bool is_first_nvshmem_kernel_in_executable = false;
+    {
+      auto [needs_nvshmem_barrier, mutex] = GetNeedsNvshmemBarrier();
+      absl::MutexLock lock(mutex);
+      auto it = needs_nvshmem_barrier->find(run_id.ToInt());
+      if (it != needs_nvshmem_barrier->end()) {
+        is_first_nvshmem_kernel_in_executable = true;
+        // Remove entry so we won't add another barrier for this module
+        // execution.
+        needs_nvshmem_barrier->erase(it);
+      }
+    }
+
+    if (is_first_nvshmem_kernel_in_executable || !use_custom_barrier) {
+      mosaic::gpu::NvshmemApi::Default().barrier_all_on_stream(
+          reinterpret_cast<cudaStream_t>(stream));
+    }
   }
-  void* args[4] = {&std::get<0>(*ctx_and_kernel), &stream, &buffers};
-  std::get<1>(*ctx_and_kernel)(args);
+  std::get<1>(*ctx_kernel_comm)(args);
+  return absl::OkStatus();
+}
+
+absl::Status MosaicGpuInitialize(ffi::RemainingArgs args,
```

Comment:
What are the semantics of this callback? Does it get called once for each custom call at XLA compile time? Is run_id an identifier of the whole HLO module or just this one custom call?

Reply:
The semantics of "initialize" is that it gets called for each HLO op of the module before each execution of the module ops.
For each run of module A:
1. Run all the "prepare" hooks of the HLO ops
2. Run all the "Initialize" hooks of the HLO ops
3. Run all the "execute" hooks of the HLO ops

The use of "run_id" was suggested to us by @ezhulenev. It represents a "logical execution" of HLO modules that can communicate via collectives. So it is similar to HLO module id representing the whole HLO module.

See explanations about the custom_barrier logic in next comment.

Reply:
What is the difference between the prepare and initialize hooks? Are perhaps one of them for compile time while the other are for runtime? Are the hooks guaranteed to be invoked in the same order in which the ops will be executed?